### PR TITLE
Added fake executor with spin loop

### PR DESCRIPTION
**Pull request by zablotchi**
## Description 

Created a fake executor as an alternative to the SuiExecutor. The fake executor simulates executing transactions by busy waiting for a predetermined amount of time.

### CODE REVIEW

File: crates/remora/src/executor/fake.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    delay: u64, // delay in ms for that this transaction simulates
+}
+pub type FakeTransactionType = Transaction<FakeExecutor>;
+
+pub type FakeExecutionResults = ExecutionResults<FakeExecutor>;
+pub struct FakeStore;
+
+impl FakeExecutionResults {
+    // todo: is there a way to make this a constant instead of a fn?
+    pub fn dummy() -> Self {
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:25]:
```suggestion
    pub fn new() -> Self {
```

File: crates/remora/src/executor/fake.rs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+pub struct FakeTransaction {
+    delay: u64, // delay in ms for that this transaction simulates
+}
+pub type FakeTransactionType = Transaction<FakeExecutor>;
+
+pub type FakeExecutionResults = ExecutionResults<FakeExecutor>;
+pub struct FakeStore;
+
+impl FakeExecutionResults {
+    // todo: is there a way to make this a constant instead of a fn?
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:24]: No need to make it constant, the compiler should do it for you. And in any case it is called only once, so there is no need to optimize it.

File: crates/remora/src/executor/fake.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    }
+}
+
+impl ExecutableTransaction for FakeTransaction {
+    fn digest(&self) -> &TransactionDigest {
+        // Return a reference to a static TransactionDigest
+        &TransactionDigest::ZERO
+    }
+
+    fn input_objects(&self) -> Vec<InputObjectKind> {
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:39]: Perhaps for a next PR (you did good to now bring all the complexity at once), but we would need to vary the number of inputs for this transaction (so we can test our resharding strategy)

File: crates/remora/src/executor/fake.rs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        Vec::new()
+    }
+}
+
+impl ObjectStore for FakeStore {
+    fn get_object(
+        &self,
+        _object_id: &ObjectID,
+    ) -> Result<Option<Object>, sui_types::storage::error::Error> {
+        unreachable!()
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:50]: Same comment here, the Fake executor will be useful to test CPU but also memory.

File: crates/remora/src/executor/fake.rs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    fn get_object_by_key(
+        &self,
+        _object_id: &ObjectID,
+        _version: VersionNumber,
+    ) -> Result<Option<Object>, sui_types::storage::error::Error> {
+        unreachable!()
+    }
+}
+
+impl ParentSync for FakeStore {
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:62]: Why do we need this? (I may be missing something)

File: crates/remora/src/executor/fake.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    fn get_latest_parent_entry_ref_deprecated(
+        &self,
+        _object_id: ObjectID,
+    ) -> SuiResult<Option<ObjectRef>> {
+        unreachable!()
+    }
+}
+
+impl ChildObjectResolver for FakeStore {
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:71]: Same question here. Are we sure we need this?

File: crates/remora/src/executor/fake.rs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        _owner: &ObjectID,
+        _receiving_object_id: &ObjectID,
+        _receive_object_at_version: SequenceNumber,
+        _epoch_id: EpochId,
+    ) -> SuiResult<Option<Object>> {
+        unimplemented!()
+    }
+}
+
+impl BackingPackageStore for FakeStore {
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:100]: I suspect that this project will not support publishing smart contract (we only load them upon booting)

File: crates/remora/src/executor/fake.rs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        _ctx: Option<std::sync::Arc<BenchmarkContext>>,
+        _store: std::sync::Arc<Self::Store>,
+        transaction: &FakeTransactionType,
+    ) -> FakeExecutionResults {
+
+        let start = Instant::now();
+        let duration = Duration::from_millis(transaction.delay);
+    
+        // Busy-wait until the specified duration has elapsed.
+        while Instant::now() - start < duration {}
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:132]: mmh it seems that the compiler optimises away this line

File: crates/remora/src/executor/fake.rs [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    }
+    
+    async fn load_state_for_shared_objects(&self) {}
+}
+
+pub fn generate_fake_transaction(delay: u64) -> FakeTransaction {
+    FakeTransaction { delay }
+}
+
+pub fn generate_fake_transactions_normal_distribution(mean: u64, std_dev: u64, num_transactions: u64) -> Vec<FakeTransaction> {
```
Comment (asonnino) [inline crates/remora/src/executor/fake.rs:155]: Should we re-use the models locaed in mock_consensus.rs?

File: crates/remora/src/executor/api.rs [hunk 10]
```diff
@@ -110,18 +110,22 @@ pub trait Executor {
     type Store: StateStore<Self::ExecutionResults>;
 
     /// Get the context for the benchmark.
-    fn context(&self) -> Arc<BenchmarkContext>;
+    fn context(&self) -> Option<Arc<BenchmarkContext>>;
```
Comment (asonnino) [inline crates/remora/src/executor/api.rs:113]: Would it be cleaner to re-define a generic BenchmarkContext?

File: crates/remora/src/executor/sui.rs [hunk 11]
```diff
@@ -323,18 +309,33 @@ impl Executor for SuiExecutor {
...
+    async fn load_state_for_shared_objects(&self) {
+        if let WorkloadType::SharedObjects { .. } = self.workload_type {
+            // import txs to assign shared-object versions
+            let (_, read_txs) = import_from_files(self.log_dir_path.clone().unwrap());
+            self.ctx
+                .validator()
+                .assigned_shared_object_versions(&read_txs)
+                .await;
+        }
+    }
```
Comment (asonnino) [inline crates/remora/src/executor/sui.rs:338]: That's nice indeed

**Review by asonnino — CHANGES_REQUESTED** [review]
I think the compiler optimises away the busy loop. Have you tried on your machinet to see if you reach close to 100% CPU utilisation?

File: crates/remora/src/executor/fake.rs (hunk 8 above)
Comment (Scofield626) [reply crates/remora/src/executor/fake.rs:132]: Here's my [code](https://gist.github.com/Scofield626/48bc1926dbd22a197573d9d7b0142ff6) for having a calibrated iterations of fake works according to the txn time we need. It works out well for my previous experiments. The only nitty thing is on line 14, which you may need a large number for txn runtime longer than 10ms.

**Comment by zablotchi** [comment]
Handled by #2

**Pull request closed without merging**